File: 02-exercise/03-solution.hs
-- Übung 02 - Programmierung TU Dresden
-- Tutor: Claas de Boer
-- Github: https://github.com/cdboer/programmierung-ss19

data Tree = Node String [Tree]

-- Aufgabe 03A
-- Geben Sie das entsprechende Element des Typs Tree an:
-- Node "Wurzel" [Node "l" [], Node "m" [Node "lu" [], Node "ru" []], Node "r" []]

-- Aufgabe 03B
level :: Int -> Tree -> [String]
level 0 (Node x _ ) = [x]
level n (Node _ ts) = concatMap (level (n-1)) ts
    where
        concatMap :: (a -> [b]) -> [a] -> [b]
        concatMap _ [] = []
        concatMap f (x:xs) = f x ++ concatMap f xs

File: 08-exercise/03-solution.pl
% Übung 08 - Programmierung TU Dresden
% Tutor: Claas de Boer
% Github: https://github.com/cdboer/programmierung-ss19

% Aufgabe 03A
edge(1, 1).
edge(1, 4).
edge(1, 2).
edge(3, 2).
edge(4, 3).

% Aufgabe 03B
path(U, U).
path(U, W) :- edge(U, V), path(V, W).

% Aufgabe 03C

%         ?- path(4, X).
% {X = 4} ?- .

%         ?- path(4, X).
%         ?- edge(4, W), path(W, X).
% {W = 3} ?- path(3, X).
% {X = 3} ?- .

%         ?- path(4, X).
%         ?- edge(4, W), path(W, X).
% {W = 3} ?- path(3, X).
%         ?- edge(3, U), path(U, X).
% {U = 2} ?- path(2, X).
% {X = 2} ?- .
